sdk/monitor/monitor-opentelemetry/src/traces/azureFnHook.ts

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { Context as AzureFnContext } from "@azure/functions";
+import {
+  context,
+  propagation,
+  ROOT_CONTEXT,
+  Context as OpenTelemetryContext,
+} from "@opentelemetry/api";
+import { Logger } from "../shared/logging";
+
+export class AzureFunctionsHook {
+  private _functionsCoreModule: any;
+  private _preInvocationHook: any;
+
+  constructor() {
+    try {
+      // TODO: Add types files when publicly available
+      this._functionsCoreModule = require("@azure/functions-core");
+      // Only v3 of Azure Functions library is supported right now. See matrix of versions here:
+      // https://github.com/Azure/azure-functions-nodejs-library
+      const funcProgModel = this._functionsCoreModule.getProgrammingModel();
+      if (funcProgModel.name === "@azure/functions" && funcProgModel.version.startsWith("3.")) {
+        this._addPreInvocationHook();
+      } else {
+        Logger.getInstance().debug(
+          `AzureFunctionsHook does not support model "${funcProgModel.name}" version "${funcProgModel.version}"`
+        );
+      }
+    } catch (error) {
+      Logger.getInstance().debug(
+        "@azure/functions-core failed to load, not running in Azure Functions"
+      );
+    }
+  }
+
+  public shutdown() {
+    if (this._preInvocationHook) {
+      this._preInvocationHook.dispose();
+      this._preInvocationHook = undefined;
+    }
+    this._functionsCoreModule = undefined;
+  }
+
+  private _addPreInvocationHook() {
```

Comment:
Nothing wrong since I see this is the same way tracking context in Azure Functions in the beta works, but I'm just wondering for my understanding why we can drop the `_addPostInvocationHook()` method when we move away from the legacy SDK.

Reply:
Pre invocation is used to maintain correlation and Post Invocation is used to automatically generate telemetry, for auto generation of telemetry we need to have further discussions, that is why is not part of this